### PR TITLE
Types for query methods

### LINKED ISSUE #236: Custom Query Methods Type Tracking
Currently there is no practical way to add the type of an custom query function to the `DocumentQuery` type so it needs to be casted

### PR DESCRIPTION
**Pull request by ggurkal**
This is the only thing I could think of for query methods types. Using mongoose's own types for `Model`'s `QueryHelpers`, which is added as the third type argument to the `getModelForClass` function.

- closes #236

### CODE REVIEW

**Comment by hasezoey** [comment]
this is more or less exactly what typegoose tries to "prevent" (making interfaces for mongoose schemas/models)

---

otherwise, the only think i would like to change is to swap the generics of `getModelForClass` to be more user friendly definable

**Comment by ggurkal** [comment]
Yes, I know and I don't really like it either but otherwise there are no types at all for people who'd like to have them.

I'll investigate it a bit further to see if it's possible have types without defining interfaces explicitly.

**Comment by hasezoey** [comment]
> I'll investigate it a bit further to see if it's possible to have types without defining interfaces explicitly.

because if we would make the functions *in the class* again, we would need an `instanceMethod` and `queryMethod` decorators and would have to filter them out again and when making it outside of the class (and class-level-decorators like currently), then there are no types automatically, so the best option is to use an interface (to define one or more query functions)
-> so i guess you dont need to search for this, because there will not be an other simpler way

**Comment by ggurkal** [comment]
I have two easier solutions/proposals and assume we have the following in both.

```ts
function findByName(...) { ... }

function findByLastname(...) { ... }

@queryMethod(findByName)
@queryMethod(findByLastname)
class MyClass {
  @prop()
  public name: string;

  @prop()
  public lastname: string;
}
```

Usage:
```ts
const MyModel = getModelForClass<MyClass, typeof MyClass, SchemaQueries>(MyClass);
```

---

1. Easier interface
```ts
// Will be placed in `types.ts`
type QueryMethod<T extends (...args: any) => any> = (...args: Parameters<T>) => ReturnType<T>;

// In model's file
interface SchemaQueries {
  findByName: QueryMethod<typeof findByName>;
  findByLastname: QueryMethod<typeof findByLastname>;
}
```

2. Defining functions in an object.
```ts
// Will be placed in `types.ts`
type QueryMethodTypes<T extends {}> = {
  [F in keyof T]: T[F] extends (...args: any) => any ? (...args: Parameters<T[F]>) => ReturnType<T[F]> : never;
}

// In model's file
const MyMethods = { findByName, findByLastname };
type SchemaQueries = QueryMethodTypes<typeof MyMethods>;
```

**Comment by hasezoey** [comment]
i think solution 1 is better and easier to understand & modify

**Comment by ggurkal** [comment]
I agree.

Do you think we should make `QueryHelpers` type the first type in `getModelForClass` ?

**Comment by hasezoey** [comment]
> Do you think we should make QueryHelpers type the first type in getModelForClass ?

yes, but this would be an breaking change (maybe even major)

File: test/tests/shouldRun.test.ts [hunk 1]
```diff
@@ -425,9 +425,17 @@ it('should add query Methods', async () => {
...
+  const metadata: QueryMethodMap = Reflect.getMetadata(DecoratorKeys.QueryMethod, QueryMethods);
+  expect(Array.from(metadata)).toEqual(
+    expect.arrayContaining([['findByName', findByName]]) &&
+    expect.arrayContaining([['findByLastname', findByLastname]])
+  );
+
+  const doc = await QueryMethodsModel.create({ name: 'hello', lastname: 'world' });
+
+  const found: DocumentType<QueryMethods>[] | null =
+    await (QueryMethodsModel.find() as any).findByName('hello').findByLastname('world').exec();
```
Comment (hasezoey) [inline test/tests/shouldRun.test.ts]: maybe remove the `as any` now that they types should work

**Comment by ggurkal** [comment]
@hasezoey are you planning to merge this PR or should we make it specific to updating the related types in `getModelForClass` and keep it until they’re ready?

**Comment by hasezoey** [comment]
> are you planning to merge this PR

i still need to check out this pr locally, and if all works, then i have some changes in mind i want to test & integrate

**Comment by hasezoey** [comment]
@ggurkal are you ok with the changes i made?

File: src/types.ts [hunk 2]
```diff
@@ -433,6 +433,9 @@ export interface IPluginsArray<T> {
  */
 export type VirtualPopulateMap = Map<string, any & VirtualOptions>;
 
+
+export type QueryMethod<T extends (...args: any) => any> = (...args: Parameters<T>) => ReturnType<T>;
```
Comment (hasezoey) [inline src/types.ts:437]: @ggurkal if you know what this does, could you add some tsdoc to it?

Comment (ggurkal) [reply src/types.ts:454]: Done

**Comment by ggurkal** [comment]
> @ggurkal are you ok with the changes i made?

Yes

**Pull request merged**